src/python/grpcio_tests/tests/observability/_observability_test.py

```diff
@@ -0,0 +1,399 @@
+        self._set_config_file(_VALID_CONFIG_TRACING_STATS)
+        with grpc_observability.GCPOpenCensusObservability(
+                exporter=self.test_exporter):
+            self._start_server()
+            self.unary_stream_call()
+
+        self.assertGreater(len(self.all_metric), 0)
+        self.assertGreater(len(self.all_span), 0)
+        self._validate_metrics(self.all_metric)
+        self._validate_spans(self.all_span)
+
+    def testRecordStreamUnary(self):
+        self._set_config_file(_VALID_CONFIG_TRACING_STATS)
+        with grpc_observability.GCPOpenCensusObservability(
+                exporter=self.test_exporter):
+            self._start_server()
+            self.stream_unary_call()
+
+        self.assertTrue(len(self.all_metric) > 0)
+        self.assertTrue(len(self.all_span) > 0)
+        self._validate_metrics(self.all_metric)
+        self._validate_spans(self.all_span)
+
+    def testRecordStreamStream(self):
+        self._set_config_file(_VALID_CONFIG_TRACING_STATS)
+        with grpc_observability.GCPOpenCensusObservability(
+                exporter=self.test_exporter):
+            self._start_server()
+            self.stream_stream_call()
+
+        self.assertGreater(len(self.all_metric), 0)
+        self.assertGreater(len(self.all_span), 0)
+        self._validate_metrics(self.all_metric)
+        self._validate_spans(self.all_span)
+
+    def testNoRecordBeforeInit(self):
+        self._set_config_file(_VALID_CONFIG_TRACING_STATS)
+        self._start_server()
+        self.unary_unary_call()
+        self.assertEqual(len(self.all_metric), 0)
+        self.assertEqual(len(self.all_span), 0)
+        self._server.stop(0)
+
+        with grpc_observability.GCPOpenCensusObservability(
+                exporter=self.test_exporter):
+            self._start_server()
+            self.unary_unary_call()
+
+        self.assertGreater(len(self.all_metric), 0)
+        self.assertGreater(len(self.all_span), 0)
+        self._validate_metrics(self.all_metric)
+        self._validate_spans(self.all_span)
+
+    def testNoRecordAfterExit(self):
+        self._set_config_file(_VALID_CONFIG_TRACING_STATS)
+        with grpc_observability.GCPOpenCensusObservability(
+                exporter=self.test_exporter):
+            self._start_server()
+            self.unary_unary_call()
+
+        self.assertGreater(len(self.all_metric), 0)
+        self.assertGreater(len(self.all_span), 0)
+        current_metric_len = len(self.all_metric)
+        current_spans_len = len(self.all_span)
+        self._validate_metrics(self.all_metric)
+        self._validate_spans(self.all_span)
+
+        self.unary_unary_call()
+        self.assertEqual(len(self.all_metric), current_metric_len)
+        self.assertEqual(len(self.all_span), current_spans_len)
+
+    def testTraceSamplingRate(self):
+        # Make 20 UnaryCall's
+        # With 50% sampling rate, we should get 5-15 traces with 98.8% probability
```

Comment:
I've confirmed this by running the tests manually. Exactly 12 failures out of 1000 runs, all caused by this test. Nicely calculated.

But 98.8% is too low. Since this is within our control, let's aim for 99.9% or higher. We can do that by either pumping up the number of calls made (which increases the test length) and/or by widening the bounds (which could increase the number of false green runs).

If the system is working (with 50% sampling), when we make $`n`$ calls, the probability that $`k`$ calls will be traced, with $`0 \leq k \leq n`$, is $`\frac{n \choose k}{2^n}`$. We can write some code to calculate this for us:

```python
import functools

@functools.cache
def factorial(n):
    if n == 0:
        return 1
    return n * factorial(n - 1)

@functools.cache
def choose(n, k):
    return factorial(n) / (factorial(k) * factorial(n - k))

def probabilty_excluding_tails(variables, tail_length):
    assert tail_length * 2 <= variables
    n = variables
    prob_failure = 0.0
    # Only sum the left tail
    for k in range(tail_length):
        prob_failure += (choose(n, k)) / (2.0 ** n)

    # Right tail is equal in probability
    prob_failure *= 2.0

    return 1.0 - prob_failure
```

This reproduces your 98.8% figure:

```python
>>> prob.probabilty_excluding_tails(20, 5)
0.9881820678710938
```

It looks like doubling the number of RPCs and keeping the bounds the same should be more than sufficient:

```python
>>> prob.probabilty_excluding_tails(40, 5)
0.9999998142975528
```

If you're worried about test duration, then you can run multiple iterations, breaking if the assertions are met. If they're not met, then continue with more RPCs but the same bounds. This will very quickly converge on 100% passing.